### PR TITLE
begins building out frontend to cs integration

### PR DESCRIPTION
**Pull request by tonytheleg**
### What this PR does
* adds logic to get/create/delete clusters using a local running CS
* removes any calls to frontend/cache.go in all CRUD ops to leverage CS instead
* adds ocm pkg for translating and instantiating CS Cluster objects and HCP Cluster objects
* Makefile and dev doc updates for changes
* Minor bug fixes:
  * Static validation was not accounting for operations on subscriptions like registration
  * utils/create.go updated for proper cluster input needed for CS cluster creation
  * database cache fix for SetClusterDoc as it was using wrong value

Jira: 
Link to demo recording: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Deployment: The deployment process was considered and addressed if required
- [ ] Testing: New code requires new unit tests.
- [ ] Documentation: Is the documentation updated? Either in the doc located in focus area, in the README or in the code itself.
- [ ] Customers: Is this change affecting customers? Is the release plan considered?

### CODE REVIEW

**Review by mjlshen — DISMISSED** [review]
Suggested changes that got implemented - was able to run Frontend/CS locally and confirm that I can create a subscription, create/get/delete an hcpopenshiftcluster through frontend while seeing it make the correct calls and enact the expected behavior in CS

File: frontend/pkg/frontend/frontend.go [hunk 1]
```diff
@@ -258,12 +261,36 @@ func (f *Frontend) ArmResourceRead(writer http.ResponseWriter, request *http.Req
 
 	// URL path is already lowercased by middleware.
 	resourceID := request.URL.Path
-	cluster, found := f.cache.GetCluster(resourceID)
-	if !found {
-		writer.WriteHeader(http.StatusNotFound)
+	subscriptionID := request.PathValue(PathSegmentSubscriptionID)
+	doc, err := f.dbClient.GetClusterDoc(ctx, resourceID, subscriptionID)
+	if err != nil {
+		if errors.Is(err, database.ErrNotFound) {
```
Comment (mbarnes) [inline frontend/pkg/frontend/frontend.go:267]: Our cluster database is still technically a cache and I'm wondering if we should account for cache misses in case the database is somehow incomplete.  Instead of failing here we could ask Cluster Service to search for the cluster with [ClustersListRequest.Search](https://pkg.go.dev/github.com/openshift-online/ocm-sdk-go@v0.1.422/clustersmgmt/v1#ClustersListRequest.Search), and update the cache if we get a hit.  (Probably would want to break this out into a new function since multiple endpoints would need to do this.)

File: frontend/pkg/frontend/frontend.go [hunk 2]
```diff
@@ -296,17 +323,55 @@ func (f *Frontend) ArmResourceCreateOrUpdate(writer http.ResponseWriter, request
...
+	resourceGroup := request.PathValue(PathSegmentResourceGroupName)
 	subscriptionID := request.PathValue(PathSegmentSubscriptionID)
 
-	cluster, updating := f.cache.GetCluster(resourceID)
-	versionedCurrentCluster := versionedInterface.NewHCPOpenShiftCluster(cluster)
+	var doc *database.HCPOpenShiftClusterDocument
+	var updating bool = true
+	doc, err = f.dbClient.GetClusterDoc(ctx, resourceID, subscriptionID)
+	if err != nil {
+		if errors.Is(err, database.ErrNotFound) {
```
Comment (mbarnes) [inline frontend/pkg/frontend/frontend.go:333]: Similar comment as above -- maybe we want to ask Cluster Service to search to make sure the cluster _really_ doesn't exist before initiating a cluster creation.

File: frontend/pkg/frontend/frontend.go [hunk 3]
```diff
@@ -399,12 +460,28 @@ func (f *Frontend) ArmResourceDelete(writer http.ResponseWriter, request *http.R
...
 	subscriptionID := request.PathValue(PathSegmentSubscriptionID)
 
-	_, found := f.cache.GetCluster(resourceID)
-	if !found {
-		writer.WriteHeader(http.StatusNotFound)
-		return
+	var doc *database.HCPOpenShiftClusterDocument
+	doc, err = f.dbClient.GetClusterDoc(ctx, resourceID, subscriptionID)
+	if err != nil {
+		if errors.Is(err, database.ErrNotFound) {
```
Comment (mbarnes) [inline frontend/pkg/frontend/frontend.go:466]: Similar comment as above -- maybe we want to ask Cluster Service to search first.

File: frontend/pkg/ocm/ocm.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+					ManagedResourceGroup:   cluster.Azure().ManagedResourceGroupName(),
+					SubnetID:               cluster.Azure().SubnetResourceID(),
+					OutboundType:           "",
+					NetworkSecurityGroupID: cluster.Azure().NetworkSecurityGroupResourceID(),
+					EtcdEncryptionSetID:    "",
+				},
+				IssuerURL: "",
+				ExternalAuth: api.ExternalAuthConfigProfile{
+					Enabled:       false,
+					ExternalAuths: []*v1.OIDCProvider{},
```
Comment (mbarnes) [inline frontend/pkg/ocm/ocm.go:84]: Was `ExternalAuth` intentionally skipped?  I think it's fine to skip it on the 1st pass because, honestly, it's a pain.  But we should leave a TO-DO comment or something similar here, and maybe even add a JIRA story to the backlog so it's not forgotten.

File: frontend/pkg/ocm/ocm.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+				},
+				Ingress: []*api.IngressProfile{},
```
Comment (mbarnes) [inline frontend/pkg/ocm/ocm.go:86]: Similar question: was `Ingress` intentionally skipped?
Ingress profiles are simple enough that we should just do it now.

**Review by mbarnes — CHANGES_REQUESTED** [review]
Had a suggestion about maybe handling cache misses in our database, so we're not mistakenly responding with "Not Found".  i.e. Check the cache for a quick answer, but still treat Cluster Service as authoritative.

Also it looks like the translation to and from Cluster Service's cluster representation has some gaps.  That's fine for now but I'd like to see a follow-up JIRA story to make sure we don't lose it.

File: frontend/pkg/ocm/ocm.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+const (
+	csCloudProvider    string = "azure"
+	csProductId        string = "aro"
+	csHypershifEnabled bool   = true
+	csMultiAzEnabled   bool   = true
+	csCCSEnabled       bool   = true
+)
+
+// ConvertCStoFrontend converts a CS Cluster object into HCPOpenShiftCluster object
+func ConvertCStoFrontend(cluster *cmv1.Cluster) (*api.HCPOpenShiftCluster, error) {
```
Comment (AldoFusterTurpin) [inline frontend/pkg/ocm/ocm.go:20]: Minor optional suggestion: it could be interesting to rename this to _ConvertCStoHCPOpenShiftCluster_ as this could potentially be used in different places than just the frontend, and we don't need to couple here the concept as the function is not coupled with anything about the frontend, it is just called from there.

File: frontend/pkg/frontend/frontend.go [hunk 7]
```diff
@@ -258,12 +261,36 @@ func (f *Frontend) ArmResourceRead(writer http.ResponseWriter, request *http.Req
...
+			writer.WriteHeader(http.StatusNoContent)
+			return
+		} else {
+			f.logger.Error(err.Error())
+			writer.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+
+	cluster, err := f.conn.ClustersMgmt().V1().Clusters().Cluster(doc.ClusterID).Get().Send()
```
Comment (AldoFusterTurpin) [inline frontend/pkg/frontend/frontend.go:278]: Maybe not for this PR and maybe not useful for the team, but I think it could be a good idea to put an interface (like ClustersController or something similar) between frontend and CS to provide a contract where the type the interface understands is _api.HCPOpenShiftCluster_.

That way, we could run the frontend without CS if we want with a simple in-memory struct that implements the interface. By doing so, we can run the frontend and try it without explicitly depending on CS details. To do that, we would move all the CS specifics that are in frontend.go to its package, and in NewFrontend() pass an interface as a parameter that gets assigned to a field of the frontend. 

Then we could use an env variable (like LOCAL_CS="true" with a default value of "false") that would instantiate a fake CS instead of the "real one", and we would control that in frontend/cmd/cmd.go where we are calling NewFrontend(). That way we can decouple frontend and CS and just focus on the contract in the frontend code.

**Review by AldoFusterTurpin — COMMENTED** [review]
I just left a minor suggestion about a function name and the idea to decouple frontend and CS moving some frontend code to a different package and creating an interface.Nothing blocking this to get merged, though. Thank you!

File: frontend/pkg/frontend/frontend.go (hunk 1 above)
Comment (tonytheleg) [reply frontend/pkg/frontend/frontend.go:267]: Talking about this with @mjlshen  and @bennerv  and @miguelsorianod  on the side. Cosmos is acting as a source of truth here because hitting CS directly without a cluster id could return multiple results back. Using something like the resource name of the cluster may not be unique enough to only get one cluster back. Sounds like CS has new support for using subscription ID and resource group name for queries that might improve that using the search method but even that may not be unique enough if customers spam create the same cluster name in same sub and rg. If they delete the cluster and recreate, it should be removed from CS but might not be right away maybe?

Maybe we leave this for now but i can create a card to make sure we revist this because its for sure something we want to get right but not sure we have all the pieces 100% today? Thoughts?

File: frontend/pkg/ocm/ocm.go (hunk 4 above)
Comment (tonytheleg) [reply frontend/pkg/ocm/ocm.go:84]: Honestly some of these were skipped because they are a bunch of work and we don't have any values just yet so i left things out. 100% we need to track updating these to ensure all objects are set. I can add a comment and create a card if that works for you

File: frontend/pkg/ocm/ocm.go (hunk 5 above)
Comment (tonytheleg) [reply frontend/pkg/ocm/ocm.go:86]: I can get this added

File: frontend/pkg/frontend/frontend.go (hunk 7 above)
Comment (tonytheleg) [reply frontend/pkg/frontend/frontend.go:278]: > That way, we could run the frontend without CS if we want with a simple in-memory struct that implements the interface.

I was saying this exact thing to Michael yesterday about this, not requiring CS would be nice. Follow up PR/Card seems like a good place, i can get one created for backlog and see what folks think

File: frontend/pkg/frontend/frontend.go (hunk 1 above)
Comment (mbarnes) [reply frontend/pkg/frontend/frontend.go:267]: I'm ok with deferring this.  Cluster Service might have their own identification scheme that's name-independent but the cluster still has to be unique to ARM, which identifies resources by subscription ID, resource group name and resource name.  So I think customers can still get name collisions at the ARM level if they're spam-creating clusters like you said.  But yes, we can revisit this after Cluster Service implements their [DDR](https://docs.google.com/document/d/19Yuu-k5IjGW1dac-tSpHDD0KTjPe3gAt_NUvHaTV9Yk/edit).

File: frontend/Makefile [hunk 8]
```diff
@@ -13,6 +13,9 @@ REGION ?= eastus
 frontend:
 	go build -o aro-hcp-frontend .
 
+run:
+	./aro-hcp-frontend --use-cache --region eastus --clusters-service-url http://localhost:8000
```
[the remark this answers is not in the document]
Comment (tonytheleg) [reply frontend/Makefile]: Updated

File: frontend/pkg/frontend/frontend.go (hunk 1 above)
Comment (tonytheleg) [reply frontend/pkg/frontend/frontend.go:267]: Will get a card created today

File: frontend/pkg/ocm/ocm.go (hunk 6 above)
Comment (mbarnes) [reply frontend/pkg/ocm/ocm.go]: Agree, and another thing to keep in mind is we'll need a similar function for node pools.

File: frontend/pkg/ocm/ocm.go (hunk 5 above)
Comment (tonytheleg) [reply frontend/pkg/ocm/ocm.go]: Ive updated this to empty values right now as it may not be as one-to-one as i hoped. in CS its a list of Ingresses that we'd have to loop through, but URL and IP dont appear to be available values, only Visibility (in the form of ListeningMethod) so this might need to be revisted. Unless i totally missed something!

File: frontend/pkg/ocm/ocm.go (hunk 6 above)
Comment (tonytheleg) [reply frontend/pkg/ocm/ocm.go]: Updated

File: frontend/pkg/frontend/middleware_validatestatic.go
Comment (mjlshen) [inline frontend/pkg/frontend/middleware_validatestatic.go:1]: I think we can revert the changes in this file since this got split out into #167

**Review by mjlshen — CHANGES_REQUESTED** [review]

**Review by mjlshen — APPROVED** [review]
